diff --git a/src/cxa_personality.cpp b/src/cxa_personality.cpp
index 63adf31..da2d9cf 100644
--- a/src/cxa_personality.cpp
+++ b/src/cxa_personality.cpp
@@ -336,6 +336,8 @@ static const void* read_target2_value(const void* ptr)
 #endif
 }
 
+#endif
+#if 0
 static const __shim_type_info*
 get_shim_type_info(uint64_t ttypeIndex, const uint8_t* classInfo,
                    uint8_t ttypeEncoding, bool native_exception,
