diff --git a/src/cxa_personality.cpp b/src/cxa_personality.cpp
index 63adf31..527d3b1 100644
--- a/src/cxa_personality.cpp
+++ b/src/cxa_personality.cpp
@@ -347,7 +347,7 @@ get_shim_type_info(uint64_t ttypeIndex, const uint8_t* classInfo,
         call_terminate(native_exception, unwind_exception);
     }
 
-    assert(ttypeEncoding == DW_EH_PE_absptr && "Unexpected TTypeEncoding");
+    assert((ttypeEncoding & 0x0f) == DW_EH_PE_absptr && "Unexpected TTypeEncoding");
     (void)ttypeEncoding;
 
     const uint8_t* ttypePtr = classInfo - ttypeIndex * sizeof(uintptr_t);
@@ -414,7 +414,7 @@ exception_spec_can_catch(int64_t specIndex, const uint8_t* classInfo,
         call_terminate(false, unwind_exception);
     }
 
-    assert(ttypeEncoding == DW_EH_PE_absptr && "Unexpected TTypeEncoding");
+    assert((ttypeEncoding & 0x0f) == DW_EH_PE_absptr && "Unexpected TTypeEncoding");
     (void)ttypeEncoding;
 
     // specIndex is negative of 1-based byte offset into classInfo;
